### PR TITLE
feat: use IPNI advertisements from the miner only

### PR DESCRIPTION
**Pull request by bajtos**
1. Resolve `minerId` to miner's PeerId, which is the same value as `Provider.ID` in the IPNI records

2. Ignore IPNI records advertised by different miners

3. Report `minerId` and `providerId` as the new measurement fields

Links:
- https://github.com/filecoin-station/roadmap/issues/65

### CODE REVIEW

File: test/integration.js [hunk 1]
```diff
@@ -1,8 +1,11 @@
 import Spark from '../lib/spark.js'
-import { test } from 'zinnia:test'
+
 import { assert, assertEquals } from 'zinnia:assert'
+import { test } from 'zinnia:test'
```
Comment (bajtos) [inline test/integration.js:4]: I used VS Code's `Organize imports` action and I'd like to keep test dependencies separated from "lib" dependencies.

File: test/integration.js [hunk 2]
```diff
@@ -16,16 +19,21 @@ test('integration', async () => {
...
+  }
   const measurementId = await spark.nextRetrieval()
   const res = await fetch(`https://api.filspark.com/measurements/${measurementId}`)
   assert(res.ok)
   const m = await res.json()
   const assertProp = (prop, expectedValue) => assertEquals(m[prop], expectedValue, prop)
 
   assertProp('cid', KNOWN_CID)
-  // TODO - spark-api does not record this field yet
-  // assertProp('indexerResult', 'OK')
```
Comment (bajtos) [inline test/integration.js:28]: Unrelated change; I am happy to do it in a different pull request if you prefer.

Comment (bajtos) [reply test/integration.js:28]: On the second thought, let's do that. See https://github.com/filecoin-station/spark/pull/57

File: lib/spark.js [hunk 3]
```diff
@@ -39,8 +41,34 @@ export default class Spark {
...
+      if (err.name === 'FilecoinRpcError') {
+        // TODO: report the error to Sentry
+        console.error('The error printed above was not expected, please report it on GitHub:')
+        console.error('https://github.com/filecoin-station/spark/issues/new')
+      } else {
+        this.#activity.onError()
+      }
+      err.reported = true
+      // Abort the check, no measurement should be recorded
+      throw err
```
Comment (juliangruber) [inline lib/spark.js:67]: I don't understand the need for this custom property `.reported` here.

If we remove the `console.error(err)` above, then we can let the outer catch handler log it. Also, the outer catch handler can call `#activity.onError()` itself.

I think the only custom error handling we need here is when `err.name === 'FilecoinRpcError'`, and in that case we can simply add a new log event. I'm proposing to refactor this catch handler like this:

```suggestion
      // There are three common error cases:
      //  1. We are offline
      //  2. The JSON RPC provider is down
      //  3. JSON RPC errors like when Miner ID is not a known actor
      // There isn't much we can do in the first two cases. We can notify the user that we are not
      // performing any jobs and wait until the problem is resolved.
      // The third case should not happen unless we made a mistake, so we want to learn about it
      if (err.name === 'FilecoinRpcError') {
        // TODO: report the error to Sentry
        console.error('Unexpected error, please report it on GitHub:')
        console.error('https://github.com/filecoin-station/spark/issues/new')
      }
      // Abort the check, no measurement should be recorded
      throw err
```

File: lib/constants.js [hunk 4]
```diff
@@ -1,3 +1,9 @@
 export const SPARK_VERSION = '1.9.0'
 export const MAX_CAR_SIZE = 200 * 1024 * 1024 // 200 MB
 export const APPROX_ROUND_LENGTH_IN_MS = 20 * 60_000 // 20 minutes
+
+export const RPC_REQUEST = new Request('https://api.node.glif.io/', {
+  headers: {
+    authorization: 'Bearer 6bbc171ebfdd78b2644602ce7463c938'
```
Comment (juliangruber) [inline lib/constants.js:7]: Can you please confirm that this is a new token, used only by SPARK?

**Review by juliangruber — CHANGES_REQUESTED** [review]

File: test/integration.js [hunk 5]
```diff
@@ -16,14 +19,20 @@ test('integration', async () => {
 
 test('retrieval check for our CID', async () => {
   const spark = new Spark()
-  spark.getRetrieval = async () => ({ cid: KNOWN_CID })
+  spark.getRetrieval = async () => ({ cid: KNOWN_CID, minerId: OUR_FAKE_MINER_ID })
+  spark.lookupMinerPeerId = async (minerId) => {
```
[the remark this answers is not in the document]
Comment (bajtos) [reply test/integration.js:23]: In this particular test, if my stubbed function isn't called, the retrieval check fails because Filecoin's RPC API call will reject the fake miner id.

The purpose of this test is not to verify how Spark works under the hood (and whether it calls all other functions we think it should call) but whether it can find the provider and retrieve the data for the CID served by our Frisbii instance.

However, if you think it's worth asserting that both functions were called, then I am fine with adding that. Let me know!

File: lib/miner-lookup.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
+import { RPC_REQUEST } from './constants.js'
+
+/**
+ * @param {string} minerId A miner actor id, e.g. `f0142637`
+ * @returns {Promise<string>} Miner's PeerId, e.g. `12D3KooWMsPmAA65yHAHgbxgh7CPkEctJHZMeM3rAvoW8CZKxtpG`
+ */
+export async function lookupMinerPeerId (minerId) {
```
[the remark this answers is not in the document]
Comment (bajtos) [reply lib/miner-lookup.js:7]: Good point. Let's rename this function to `getMinerPeerId`.

File: lib/constants.js (hunk 4 above)
Comment (bajtos) [reply lib/constants.js:7]: Yes, it's a new token associated with my PL email address, used only here.

File: lib/spark.js [hunk 7]
```diff
@@ -16,6 +17,7 @@ export default class Spark {
 
   constructor ({ fetch = globalThis.fetch } = {}) {
     this.#fetch = fetch
+    this.lookupMinerPeerId = lookupMinerPeerId
```
[the remark this answers is not in the document]
Comment (bajtos) [reply lib/spark.js]: I am fine either way.

The downside of your proposal is that I have to rename the imported implementation of `getMinerPeerId`, otherwise I get the following error:


```
error: ReferenceError: Cannot access 'getMinerPeerId' before initialization
    getMinerPeerId = getMinerPeerId,
                     ^
    at new Spark (file:///Users/bajtos/src/filstation/spark/lib/spark.js:21:22)
```

[the remark this answers is not in the document]
Comment (bajtos) [reply lib/spark.js]: See 7378e8d

[the remark this answers is not in the document]
Comment (bajtos) [reply lib/spark.js]: See [7378e8d](https://github.com/filecoin-station/spark/pull/55/commits/7378e8d775cce52754615260a0a60f89d921c65b)

File: lib/spark.js (hunk 3 above)
Comment (bajtos) [reply lib/spark.js:71]: Let's discuss.

In my proposed implementation:
- FilecoinRpcError does not trigger `this.#activity.onError()`, therefore Spark stays in "online" mode
- We print the call to report the issue after the error stack.

In your proposal:
- FilecoinRpcError triggers `this.#activity.onError()` and puts Spark to offline mode. 
- The call to report the issue is printed before the error to report. (I guess that's fine, we just need to tweak the message.)

Here is the important question we need to answer: How should we handle the case when we cannot get a miner's Peer ID because Filecoin RPC returns an error? 
- This can be either because we are calling the RPC API incorrectly or because the miner ID is not a valid one. (I suppose we could also receive this error if the RPC API server has an internal problem; it can process the JSON-RPC part, but the called RPC method fails.)
- Do we want to indicate offline mode? The next retrieval check will likely be able fetch the round details from spark-api, which means Spark goes back to online mode, only to go offline again when the JSON RPC call fails.

File: lib/spark.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (juliangruber) [reply lib/spark.js]: Yeah that's unfortunate, but, to me, conceptionally it's cleaner to rename a variable than to use two methods for mocking

File: test/integration.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (juliangruber) [reply test/integration.js]: This was also my understanding when I wrote tests like this, but through your reviews I understood that it's clearer and safer to assert the function calls. If in the future a condition is introduced which skips `lookupMinerPeerId()`, then this test will still pass.

However,iIn my previous review I missed that this was an integration test and not a unit test. Please choose the testing method that you see fits best.

File: lib/spark.js (hunk 3 above)
Comment (juliangruber) [reply lib/spark.js:71]: Ah understood! I thought the purpose of `err.reported` was to prevent double log lines, now I see it's also to prevent reporting it to the activity state handler.

Could we be more explicit about this?
1. Rename to `error.setActivityToOffline = false`
2. Let the outer error handler handle logging, to be more consistent with other errors

**Review by juliangruber — CHANGES_REQUESTED** [review]

File: test/integration.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (bajtos) [reply test/integration.js]: > push the function calls into an array, and then assert this function was called

added in 0a48986

File: lib/spark.js (hunk 3 above)
Comment (bajtos) [reply lib/spark.js]: See db3b2ae
